File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
#백준-회의실배정,순회강연과 비슷

from heapq import heappush, heappop

def solution(book_time):
    answer = 1
    
    # "HH:MM" → HH * 60 + MM
    time = [(int(start[:2]) * 60 + int(start[3:]), 
             int(end[:2]) * 60 + int(end[3:])) 
            for start, end in book_time]
    time.sort()
    
    heap = []
    for start, end in time:
        if not heap:
            heappush(heap,end+10)
            continue
        if heap[0] <= start:
            heappop(heap)
        else:
            answer += 1
        heappush(heap,end+10)
    
    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

public class Solution {
    public int solution(String[][] book_time) {
        List<int[]> times = new ArrayList<>();
        for (String[] time : book_time) {
            int start = convert(time[0]);
            int end = convert(time[1]) + 10;
            times.add(new int[]{start, end});
        }

        times.sort(Comparator.comparingInt(a -> a[0]));
        Queue<Integer> room = new PriorityQueue<>();
        for (int[] time : times) {
            int start = time[0];
            int end = time[1];
            if (!room.isEmpty() && room.peek() <= start) {
                    room.poll();
            }
            room.add(end);
        }
        return room.size();
    }
    private int convert(String time) {
        String[] parts = time.split(":");
        int hours = Integer.parseInt(parts[0]);
        int minutes = Integer.parseInt(parts[1]);
        return hours * 60 + minutes;
    }

}
